compiler-user-guide/src/developing-hardware/primitives.md

```diff
@@ -0,0 +1,397 @@
+
+``` yaml
+BlackBox:
+  name: Clash.Explicit.BlockRam.blockRam#
+  kind: Declaration
+  type: |-
+    blockRam#
+      :: ( KnownDomain dom        ARG[0]
+         , HasCallStack  --       ARG[1]
+         , NFDataX a )   --       ARG[2]
+      => Clock dom       -- clk,  ARG[3]
+      -> Enable dom      -- en,   ARG[4]
+      -> Vec n a         -- init, ARG[5]
+      -> Signal dom Int  -- rd,   ARG[6]
+      -> Signal dom Bool -- wren, ARG[7]
+      -> Signal dom Int  -- wr,   ARG[8]
+      -> Signal dom a    -- din,  ARG[9]
+      -> Signal dom a
+  template: |-
+    -- blockRam begin
+    ~GENSYM[~RESULT_blockRam][1] : block
+      signal ~GENSYM[~RESULT_RAM][2] : ~TYP[5] := ~CONST[5];
+      signal ~GENSYM[rd][4]  : integer range 0 to ~LENGTH[~TYP[5]] - 1;
+      signal ~GENSYM[wr][5]  : integer range 0 to ~LENGTH[~TYP[5]] - 1;
+    begin
+      ~SYM[4] <= to_integer(~ARG[6])
+      -- pragma translate_off
+                    mod ~LENGTH[~TYP[5]]
+      -- pragma translate_on
+                    ;
+      ~SYM[5] <= to_integer(~ARG[8])
+      -- pragma translate_off
+                    mod ~LENGTH[~TYP[5]]
+      -- pragma translate_on
+                    ;
+    ~IF ~VIVADO ~THEN
+      ~SYM[6] : process(~ARG[3])
+      begin
+        if ~IF~ACTIVEEDGE[Rising][0]~THENrising_edge~ELSEfalling_edge~FI(~ARG[3]) then
+          if ~ARG[7] ~IF ~ISACTIVEENABLE[4] ~THEN and ~ARG[4] ~ELSE ~FI then
+            ~SYM[2](~SYM[5]) <= ~TOBV[~ARG[9]][~TYP[9]];
+          end if;
+          ~RESULT <= fromSLV(~SYM[2](~SYM[4]))
+          -- pragma translate_off
+          after 1 ps
+          -- pragma translate_on
+          ;
+        end if;
+      end process; ~ELSE
+      ~SYM[6] : process(~ARG[3])
+      begin
+        if ~IF~ACTIVEEDGE[Rising][0]~THENrising_edge~ELSEfalling_edge~FI(~ARG[3]) then
+          if ~ARG[7] ~IF ~ISACTIVEENABLE[4] ~THEN and ~ARG[4] ~ELSE ~FI then
+            ~SYM[2](~SYM[5]) <= ~ARG[9];
+          end if;
+          ~RESULT <= ~SYM[2](~SYM[4])
+          -- pragma translate_off
+          after 1 ps
+          -- pragma translate_on
+          ;
+        end if;
+      end process; ~FI
+    end block;
+    --end blockRam
+```
+
+Again, the `name` of the primitive is the fully qualified name of the function you are creating the primitive for.
+Because we are creating a _declaration_ primitive the _kind_ must be set to `Declaration`.
+Instead of discussing what the individual template holes mean in the above context, we will instead just give a general listing of the available template holes:
+
+* `~RESULT`: Signal to which the result of a primitive must be assigned
+  to. NB: Only used in a _declaration_ primitive.
+* `~ARG[N]`: `(N+1)`'th argument to the function.
+* `~CONST[N]`: `(N+1)`'th argument to the function. Like `~ARG`, but Clash will try to reduce this to a literal, even if it would otherwise consider it too expensive.
+  And if Clash fails to reduce this argument to a literal it will produce an error.
```

Comment:
```suggestion
* `~CONST[N]`: `(N+1)`'th argument to the function.
  Like `~ARG`, but Clash will try to reduce this to a literal, even if it would otherwise consider it too expensive.
  And if Clash fails to reduce this argument to a literal it will produce an error.
```